### PR TITLE
Cleanup uses of Array::data_ref (#3880)

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #3880

# Rationale for this change
 

The first of likely multiple PRs to gradually remove the direct usage of `ArrayData::data_ref` from the kernels

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-ord/src/sort.rs [hunk 1]
```diff
@@ -1159,8 +1159,13 @@ pub struct LexicographicalComparator<'a> {
 impl LexicographicalComparator<'_> {
     /// lexicographically compare values at the wrapped columns with given indices.
     pub fn compare(&self, a_idx: usize, b_idx: usize) -> Ordering {
-        for (data, comparator, sort_option) in &self.compare_items {
-            match (data.is_valid(a_idx), data.is_valid(b_idx)) {
+        for (nulls, comparator, sort_option) in &self.compare_items {
```
Comment (alamb) [inline arrow-ord/src/sort.rs:1162]: I checked at this seems to simply pull the `match` up a level https://docs.rs/arrow-data/35.0.0/src/arrow_data/data/mod.rs.html#384-389 (as in there is no more dynamic dispatch from what I can see)

File: arrow-arith/src/boolean.rs [hunk 2]
```diff
@@ -381,29 +313,12 @@ pub fn not(left: &BooleanArray) -> Result<BooleanArray, ArrowError> {
...
-
-    let output = match input.data_ref().nulls() {
-        None => {
-            let len_bytes = ceil(len, 8);
-            MutableBuffer::from_len_zeroed(len_bytes).into()
-        }
-        Some(nulls) => buffer_unary_not(nulls.buffer(), nulls.offset(), nulls.len()),
+    let values = match input.nulls() {
+        None => NullBuffer::new_null(input.len()).into_inner(),
+        Some(nulls) => !nulls.inner(),
```
Comment (alamb) [inline arrow-arith/src/boolean.rs:318]: these certainly do look very nice now 👌

**Review by alamb — APPROVED** [review]
I think we should run the benchmarks on this PR prior to merging 

Overall though I think this is really nice progress

**Comment by tustvold** [comment]
The benchmarks do not show any major regressions, with most kernels unaffected. The boolean kernels seem to show a consistent 10% regression on some systems and a consistent 10% improvement on others, not entirely sure what is going on here but given we're talking 10s of nanoseconds I think it is probably just noise, and certainly not relevant to realistic workloads.

**Pull request merged**